Adding/removing Ports when bonds are removed/added

summeraz opened the pull request:
This PR changes the behavior of `Ports` with respect to bond formation/removal to resolve issues #382 and #383.

In our prior functionality when two `Ports` were connected using `force_overlap` their `used` attributed was changed to `True` but the `Port` objects were not destroyed. In some cases this could lead to undesirable behavior as noted in #382. This PR changes this functionality such that if two `Ports` are connected by `force_overlap` and a bond is created between the anchor particles (which is the default behavior) then the two `Ports` are removed entirely. Additionally, if a bond is removed between two `Particles` then two new `Ports` are created, each anchored to one of the two `Particles` previously involved in the bond, along the previous bond vector.

With the updated code, `Port` behavior now functions the same way on all `Compounds` regardless of how they were constructed. For example, if I create a hexane using an atom-by-atom approach in mBuild or if I load a hexane molecule from a PDB structure file, in either case I could remove a terminal hydrogen and be left with an available `Port` to add a functional group.

The only issue I see with the change in this PR is that `Port` names are lost when bonds are created (as the `Ports` themselves are destroyed); however, I cannot think of a use-case where we would want to retain these names, and I think the universal functionality this yields for all mBuild `Compounds` regardless of how they were created is a better end result. The new `Ports` that are created when bonds are removed are provided a generic name `'port[$]'`.

Something that we might want to tweak is that the `Ports` that are added when a bond is removed are currently added to the parents of the `Particles` the bond was removed between. I'm not sure if this is the best way to handle this or if the `Ports` should be added the hierarchy root or perhaps the `Compound` that `remove_bond` was called on.

summeraz commented:
Before merging I plan to create some additional unit tests to try to cover any edge cases the new behavior could cause issues with.

summeraz commented on mbuild/compound.py:
```diff
@@ -871,10 +873,23 @@ def remove_bond(self, particle_pair):
+            return
+        distance = np.linalg.norm(bond_vector)
+        particle_pair[0].parent.add(Port(anchor=particle_pair[0],
+                                         orientation=-deepcopy(bond_vector),
```
I'll go back and check. In one of my test cases `bond_vector` variable was getting changed after creating the first `Port`, but it would probably be best to identify and fix that issue rather than using `deepcopy` here as a band-aid.

ctk3b replied:
Any luck hunting this down?

ctk3b commented on mbuild/compound.py:
```diff
@@ -999,7 +999,7 @@ def center(self):
 
         """
-        if self.xyz.any():
+        if len(self.xyz) != 0:
```
These are pretty different checks. What prompted this change?

summeraz replied:
I'm still in the middle of working on this PR. This was just a temporary change (that will be fixed when #386 is merged).

summeraz commented:
I'm still running into some strange issues when the bonds that are removed are directly along the _y_ axis. Will try to get a minimal example posted later today to see if anyone has any insight.

summeraz commented:
Okay, so I ended up making several changes that I'll detail here...

It turns out that `Ports` were not being created properly, or at least how I believe they are intended to be created. Since `Ports` actually feature two subports, these subports should be created such that the positions of the `'up'` and `'down'` subports are rotated 180 degrees relative to each other (this was being done correctly) and that for two `Ports` oriented along antiparallel vectors the `'up'` and `'down'` subports of these two `Ports` feature equivalent positions with reversed labels (i.e. the `'up'` subport of a `Port` oriented in one direction will have the same xyz positions as the `'down'` subport of a `Port` oriented in the opposite direction). This second condition was not being satisfied properly in the current `Port` constructor. The issue appears to have been isolated to cases where the _z_ coordinate of the orientation vector was < 0. I could not figure out the proper rotations to satisfy this case, so as a bit of a trick, in cases where the _z_ coordinate of the orientation vector is less than zero, the subports are created as if the orientation vector was in the opoosite direction and the labels are simply switched, yielding the desired result.

I've added several unit tests to verify the `Port` creation is now being handled in the expected manner regardless of the `orientation` vector provided. I have also added several tests two ensure that when two `Compounds` are disconnected (i.e. the bond between them is removed) and reconnected (calling `force_overlap`) that the locations of the particles remain unchanged, regardless of the orientation of the bond vector.

I have also fixed a bug in `Port` constructor where `np.array_equal` was being used to test for specific `orientation` cases rather than `np.allclose`.

The docstring of the `Port` constructor has been updated.

`[0, 1, 0]` is now the default `orientation` vector, rather than `None` (since we were changing `orientation` to `[0, 1, 0]` in `__init__` anyway - this just cleans up the code a bit).

`Port`s are now rotated 180 degrees about _z_ instead of _x_ if `orientation` is [0, -1, 0]. The prior approach did not satisfy the expected behavior detailed above.

Obviously changing the `Port` creation is a big deal since this could alter existing recipes used to create `Compound`s. All of the unit tests pass, but not many of them actually test particle positions, most test connectivity. I think it would be a good idea to have a test where we have a `Compound` we've saved to a file (e.g. MOL2 or PDB, we could throw this in `utils`), and in the test we recreate this file and make sure the particle positions haven't changed. I can work on setting this up.

summeraz commented:
Also forgot to mention, thanks to @justinGilmer for helping with this

ctk3b commented on mbuild/port.py:
```diff
@@ -28,7 +34,7 @@ class Port(Compound):
 
     """
-    def __init__(self, anchor=None, orientation=None, separation=0):
+    def __init__(self, anchor=None, orientation=[0, 1, 0], separation=0):
```
This is a [mutable default arg](https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html). Could you change this to:

```python
if orientation is None: 
    orientation = [0, 1, 0]
```

summeraz replied:
Good call. That's how we had this originally, and I couldn't remember why, but this makes sense. I'll change it back.

summeraz commented:
Also, just to make note so I don't forget, I am also planning to remove the `used` attribute for `Port` as part of the PR, since this should always be `False` now that ports are removed when bonds are created. I'll try to take care of this later today.

ctk3b commented on mbuild/port.py:
```diff
@@ -45,19 +55,32 @@ def __init__(self, anchor=None, orientation=None, separation=0):
+        orientation and the labels are switched manually.
+        '''
+        swap_labels = False
+        if orientation[2] < 0:
```
This seems really strange and if at all possible I would like to understand (and fix) the root of the problem instead of circumventing it. Why is this specific to the z-dimension? Does this have to do with the rotation above? `down.rotate(np.pi, [0, 0, 1])`

summeraz replied:
Yes it is definitely strange, and I agree we should probably figure out the actual problem. I don't think it has to do with `down.rotate(np.pi, [0, 0, 1])`, since all this is doing is orienting the `down` subport in the opposite direction of `up`, prior to aligning along the `orientation` vector. But yes, it is very strange the problem is only specific to the -_z_ dimension. I've spent a decent amount of time trying to figure this out and have still not quite been able to wrap my head around it, but I will try to attach a Jupyter notebook sometime this weekend that details and visualizes the issue.

justinGilmer replied:
I haven't tested this yet, but after some discussions yesterday and last night, I think the issue could be related to the z-direction due to the way we are calculating the cross product. We always calculate it in relation to the default direction and the orientation. We then always rotate using the `rotate` method. There must be some consistent method in which we are doing this rotation that makes the z-direction the only direction that must be flipped.

ctk3b commented on mbuild/tests/base_test.py:
```diff
@@ -113,20 +119,19 @@ def benzene_from_parts(self):
 
-        ch.add(mb.Port(anchor=ch[0]), 'b')
-        mb.translate(ch['b'], [0, 0.07, 0]) 
+        ch.add(mb.Port(anchor=ch[0], separation=0.07), 'b')
```
👍

ctk3b commented on mbuild/tests/test_rigid.py:
```diff
@@ -68,12 +70,13 @@ def test_rigid_from_parts2(self, rigid_ch):
 
         for _ in range(5):
-            ch_new = mb.clone(rigid_ch)
+            ch_new = mb.clone(rigid_ch_copy)
```
You adopted this pattern in a couple of places about as well - I don't feel strongly about this so ¯\\\_(ツ)\_/¯

summeraz replied:
Good call. Your suggestion would be a bit cleaner, so I'll work on changing this.

summeraz commented on mbuild/utils/geometry.py:
```diff
@@ -0,0 +1,25 @@
+
+from mbuild.coordinate_transform import angle
+
+def calc_dihedral(point1, point2, point3, point4):
```
I had this thought as well. Actually originally I thought that perhaps I should have put `calc_dihedral` in `coordinate_transform.py` instead of `utils`. I had primarily been using this function for debugging and as it turns out the unit tests I wrote that rely on this function could actually probably be comparing the `xyz` positions of the `Compound` instead of the FCCF dihedral angle (making sure that the positions are unchanged when bonds are broken and reformed). However, the reason I kept the dihedral angle in the test is that if this code was to get broken, looking at the change in the dihedral angle rather than the full `xyz` coordinates makes it slightly easier to debug. 

So I'm not sure what the best solution is. Should we put `angle` in `geometry.py` in `utils`, or should we just move `calc_dihedral` to `coordinate_transform.py` (and rename to `dihedral_angle` as you've suggested)?

summeraz commented:
Here is a Jupyter notebook that contains some simple tests of `Port` orientation. The folder contains two copies, one that I've run on the current master and the other using this feature branch. I think this clearly demonstrates the problem and how the changes I've made with this branch are able to fix it. However, I'm still working on trying to figure out why these changes work...(e.g. why the orientation gets messed up only in -_z_ )

[simple-port-tests.tar.gz](https://github.com/mosdef-hub/mbuild/files/1488482/simple-port-tests.tar.gz)

summeraz commented:
Okay, I have this working now without the funky label switching during `Port` instantiation when the _z_ coordinate of the orientation vector is negative. I still can't really figure out why the old `Port.__init__` didn't create these properly however...

ctk3b commented on mbuild/port.py:
```diff
@@ -43,25 +54,21 @@ def __init__(self, anchor=None, orientation=None, separation=0):
-                np.asarray(default_direction), unit_vector(np.asarray(orientation))):
-            pass
+        if np.allclose(
+                np.asarray(default_direction), unit_vector(-orientation)):
```
Nit: let's cast `default_direction` to an array at the top of the function.

ctk3b commented on mbuild/tests/test_utils.py:
```diff
@@ -9,3 +11,8 @@ def test_assert_port_exists(self, ch2):
+    def test_structure_reproducibility(self, alkane_monolayer):
+        filename = 'monolayer-tmp.pdb'
+        alkane_monolayer.save(filename)
+        assert filecmp.cmp(get_fn('monolayer.pdb'), filename)
```
Excellent regression test!

summeraz replied:
Thanks. Ideally we would also be checking to make sure the topology is reproduced as well (not as relevant for this PR), but it seems the ordering of bonds, angles, etc. is lost in ParmEd so a simple file comparison won't work.

summeraz replied:
Actually it looks like the `filecmp` comparison won't work when comparing files generated on Mac or Linus with those on Windows (which is why AppVeyor is failing). @justinGilmer has suggested using `difflib` which looks like it should do the trick, so I'll update the test tonight.

ctk3b replied:
Something that I've wanted to do for a while is write a proper [__eq__ function for compound](https://github.com/mosdef-hub/mbuild/issues/64) which would let you compare the actual objects created in memory rather than the .pdb file. Clearly I've been pushing that off for quite some time based on the reassigned milestones 😬 

Not sure if that's necessarily better for this particular test but would be useful in general.
